File: sdk/python/ai_runtime/snapshot/store.py
"""
ai_runtime.snapshot.store
~~~~~~~~~~~~~~~~~~~~~~~~~~
SnapshotStore — the core snapshot/restore engine.

Architecture:
  ┌─────────────────────────────────────┐
  │   SnapshotStore(project_root)       │
  │                                     │
  │   .capture(patch_id)                │  Write lock → tarball → SQLite
  │       → SnapshotHandle              │
  │                                     │
  │   .restore(handle)                  │  Write lock → extract → SQLite update
  │                                     │
  │   .commit(handle)                   │  Mark status='committed' in SQLite
  │                                     │
  │   .list_snapshots()                 │  Read SQLite (no lock needed)
  │                                     │
  │   .prune(keep=10)                   │  Delete old archives + SQLite rows
  └─────────────────────────────────────┘

Storage layout (under project_root/.ai_runtime/):
  snapshots/                  ← tar.gz archives
      <snapshot_id>.tar.gz
  snapshots.db                ← SQLite index (WAL mode)
  snapshot.lock               ← advisory write lock file

SQLite WAL mode:
  WAL (Write-Ahead Log) allows concurrent readers while a write is in progress.
  Combined with the advisory file lock, this ensures snapshot_id is the
  canonical truth and no two agents corrupt the same archive simultaneously.

Snapshot target SLO (from architecture doc):
  - Capture time  < 100ms for projects < 50MB (source files only)
  - Archive size  < 10% of raw source size (gzip compression)
  - Restore time  < 500ms
"""

from __future__ import annotations

import os
import sqlite3
import tarfile
import tempfile
import time
import uuid
from pathlib import Path
from typing import Optional

from .gitignore import collect_source_files
from .lock import project_write_lock
from .._types import SnapshotHandle

# ─────────────────────────────────────────────────────────────────────────────
# SQL
# ─────────────────────────────────────────────────────────────────────────────

_DDL = """
CREATE TABLE IF NOT EXISTS snapshots (
    id                  TEXT PRIMARY KEY,
    patch_id            TEXT NOT NULL,
    project_root        TEXT NOT NULL,
    archive_path        TEXT NOT NULL,
    created_at          REAL NOT NULL,
    status              TEXT NOT NULL DEFAULT 'pending',
    archive_size_bytes  INTEGER NOT NULL DEFAULT 0,
    file_count          INTEGER NOT NULL DEFAULT 0
);

CREATE INDEX IF NOT EXISTS idx_snapshots_patch_id
    ON snapshots (patch_id);

CREATE INDEX IF NOT EXISTS idx_snapshots_created_at
    ON snapshots (created_at DESC);
"""


# ─────────────────────────────────────────────────────────────────────────────
# SnapshotStore
# ─────────────────────────────────────────────────────────────────────────────

class SnapshotStore:
    """
    Filesystem snapshot engine for the AI-Safe runtime.

    One instance per project root. Thread-safe via file-level advisory locking.
    Multiple SnapshotStore instances (or even processes) pointing at the same
    project_root will correctly serialize write operations.

    Args:
        project_root: Absolute (or resolvable) path to the project directory.
        store_subdir:  Sub-directory for all runtime data (default '.ai_runtime').
    """

    def __init__(
        self,
        project_root: str | Path,
        store_subdir: str = ".ai_runtime",
    ) -> None:
        self.project_root = Path(project_root).resolve()
        self._store_dir   = self.project_root / store_subdir
        self._archive_dir = self._store_dir / "snapshots"
        self._db_path     = self._store_dir / "snapshots.db"
        self._store_dir.mkdir(parents=True, exist_ok=True)
        self._archive_dir.mkdir(parents=True, exist_ok=True)
        self._init_db()

    # ── Public API ─────────────────────────────────────────────────────────

    def capture(self, patch_id: str = "") -> SnapshotHandle:
        """
        Capture the current project state into a compressed archive.

        Steps (under write lock):
          1. Collect source files (gitignore-aware)
          2. Write to a temporary tar.gz, then atomic-rename into archive_dir
          3. Record in SQLite with status='pending'

        Args:
            patch_id: ID of the patch this snapshot is being taken for.

        Returns:
            SnapshotHandle with path pointing to the archive.

        Raises:
            OSError:      If the project cannot be read or store cannot be written.
            TimeoutError: If the write lock cannot be acquired within 10s.
        """
        snap_id = str(uuid.uuid4())
        archive_path = self._archive_dir / f"{snap_id}.tar.gz"

        with project_write_lock(self.project_root):
            file_count, size_bytes = self._write_archive(archive_path)
            self._db_insert(
                snap_id=snap_id,
                patch_id=patch_id,
                archive_path=str(archive_path),
                created_at=time.time(),
                status="pending",
                size_bytes=size_bytes,
                file_count=file_count,
            )

        return SnapshotHandle(
            snapshot_id=snap_id,
            project_root=str(self.project_root),
            patch_id=patch_id,
            path=str(archive_path),
            status="pending",
            created_at=time.time(),
            archive_size_bytes=size_bytes,
        )

    def restore(self, handle: SnapshotHandle) -> None:
        """
        Restore project to the state in `handle`.

        Steps (under write lock):
          1. Verify archive exists
          2. Extract tar.gz over project_root (overwriting changed files)
          3. Update SQLite status to 'rolled_back'

        Args:
            handle: SnapshotHandle returned by a prior capture() call.

        Raises:
            FileNotFoundError: If the archive has been pruned or never written.
            TimeoutError:      If the write lock cannot be acquired.
        """
        if not handle.path or not Path(handle.path).exists():
            raise FileNotFoundError(
                f"Snapshot archive not found: {handle.path!r}. "
                "It may have been pruned or the snapshot was never fully committed."
            )

        with project_write_lock(self.project_root):
            self._extract_archive(Path(handle.path))
            self._db_update_status(handle.snapshot_id, "rolled_back")

        handle.status = "rolled_back"

    def commit(self, handle: SnapshotHandle) -> None:
        """
        Mark a snapshot as 'committed' (patch applied successfully, no rollback needed).

        Does NOT delete the archive — that is done by prune().

        Args:
            handle: SnapshotHandle to commit.
        """
        self._db_update_status(handle.snapshot_id, "committed")
        handle.status = "committed"

    def list_snapshots(self, limit: int = 50) -> list[dict]:
        """
        Return metadata for the most recent snapshots (no lock required).

        Args:
            limit: Max rows to return (default 50).

        Returns:
            List of dicts with keys: id, patch_id, status, created_at,
            archive_size_bytes, file_count, archive_path.
        """
        with self._connect() as con:
            rows = con.execute(
                "SELECT id, patch_id, status, created_at, "
                "archive_size_bytes, file_count, archive_path "
                "FROM snapshots "
                "ORDER BY created_at DESC "
                "LIMIT ?",
                (limit,),
            ).fetchall()

        return [
            {
                "id":                  r[0],
                "patch_id":            r[1],
                "status":              r[2],
                "created_at":          r[3],
                "archive_size_bytes":  r[4],
                "file_count":          r[5],
                "archive_path":        r[6],
            }
            for r in rows
        ]

    def prune(self, keep: int = 10) -> int:
        """
        Delete old 'committed' or 'rolled_back' snapshots, keeping the `keep`
        most recent ones. Acquires write lock to prevent race with capture.

        Args:
            keep: Number of recent snapshots to retain.

        Returns:
            Number of archives deleted.
        """
        with project_write_lock(self.project_root):
            return self._prune_locked(keep)

    def load(self, snapshot_id: str) -> Optional[SnapshotHandle]:
        """
        Load a SnapshotHandle from the SQLite index by ID.

        Returns None if not found.
        """
        with self._connect() as con:
            row = con.execute(
                "SELECT id, patch_id, project_root, archive_path, "
                "created_at, status, archive_size_bytes "
                "FROM snapshots WHERE id = ?",
                (snapshot_id,),
            ).fetchone()

        if row is None:
            return None

        return SnapshotHandle(
            snapshot_id=row[0],
            patch_id=row[1],
            project_root=row[2],
            path=row[3],
            created_at=row[4],
            status=row[5],
            archive_size_bytes=row[6],
        )

    # ── Private helpers ────────────────────────────────────────────────────

    def _write_archive(self, dest: Path) -> tuple[int, int]:
        """
        Write a compressed tar archive of source files to `dest`.
        Uses a temp file + atomic rename so a partial write never leaves a
        corrupt archive at the final path.

        Returns:
            (file_count, compressed_size_bytes)
        """
        tmp_fd, tmp_path = tempfile.mkstemp(
            suffix=".tar.gz.tmp",
            dir=self._archive_dir,
        )
        os.close(tmp_fd)

        file_count = 0
        try:
            with tarfile.open(tmp_path, "w:gz", compresslevel=6) as tar:
                for src in collect_source_files(self.project_root):
                    arcname = src.relative_to(self.project_root).as_posix()
                    tar.add(str(src), arcname=arcname)
                    file_count += 1

            # Atomic rename — safe on both Windows (Python 3.3+) and Unix
            os.replace(tmp_path, str(dest))
        except Exception:
            try:
                os.unlink(tmp_path)
            except FileNotFoundError:
                pass
            raise

        size_bytes = dest.stat().st_size if dest.exists() else 0
        return file_count, size_bytes

    def _extract_archive(self, archive: Path) -> None:
        """
        Extract `archive` over project_root.
        On Windows, individual file errors (locked files) are caught and reported
        collectively rather than aborting mid-restore.
        """
        errors: list[str] = []
        with tarfile.open(str(archive), "r:gz") as tar:
            for member in tar.getmembers():
                dest = self.project_root / member.name
                try:
                    dest.parent.mkdir(parents=True, exist_ok=True)
                    if member.isfile():
                        with tar.extractfile(member) as src_f:  # type: ignore[union-attr]
                            dest.write_bytes(src_f.read())
                except (OSError, PermissionError) as e:
                    errors.append(f"{member.name}: {e}")

        if errors:
            raise OSError(
                f"Restore completed with {len(errors)} file error(s):\n"
                + "\n".join(errors[:10])
            )

    def _init_db(self) -> None:
        with self._connect() as con:
            con.executescript(_DDL)

    def _connect(self) -> sqlite3.Connection:
        """Open a connection with WAL mode and sensible defaults."""
        con = sqlite3.connect(str(self._db_path), timeout=15.0)
        con.execute("PRAGMA journal_mode=WAL")
        con.execute("PRAGMA foreign_keys=ON")
        con.execute("PRAGMA synchronous=NORMAL")
        return con

    def _db_insert(
        self,
        snap_id: str,
        patch_id: str,
        archive_path: str,
        created_at: float,
        status: str,
        size_bytes: int,
        file_count: int,
    ) -> None:
        with self._connect() as con:
            con.execute(
                "INSERT INTO snapshots "
                "(id, patch_id, project_root, archive_path, created_at, "
                " status, archive_size_bytes, file_count) "
                "VALUES (?, ?, ?, ?, ?, ?, ?, ?)",
                (snap_id, patch_id, str(self.project_root),
                 archive_path, created_at, status, size_bytes, file_count),
            )
            con.commit()

    def _db_update_status(self, snapshot_id: str, status: str) -> None:
        with self._connect() as con:
            con.execute(
                "UPDATE snapshots SET status=? WHERE id=?",
                (status, snapshot_id),
            )
            con.commit()

    def _prune_locked(self, keep: int) -> int:
        """Must be called while holding the write lock."""
        with self._connect() as con:
            # Find rows to delete (oldest committed/rolled_back beyond `keep`)
            rows = con.execute(
                "SELECT id, archive_path FROM snapshots "
                "WHERE status IN ('committed','rolled_back') "
                "ORDER BY created_at DESC "
                "LIMIT -1 OFFSET ?",
                (keep,),
            ).fetchall()

            deleted = 0
            for snap_id, archive_path in rows:
                try:
                    Path(archive_path).unlink(missing_ok=True)
                except OSError:
                    pass
                con.execute("DELETE FROM snapshots WHERE id=?", (snap_id,))
                deleted += 1

            con.commit()
        return deleted

File: sdk/node/src/snapshot/store.ts
import fs from 'fs';
import path from 'path';
import crypto from 'crypto';
import * as tar from 'tar';
import properLockfile from 'proper-lockfile';
import { collectSourceFiles } from './gitignore';
import { SnapshotHandle } from '../types';

interface SnapshotRecord extends SnapshotHandle {
    file_manifest: string[];
}

export class SnapshotStore {
    private projectRoot: string;
    private storeDir: string;
    private archiveDir: string;
    private indexPath: string;
    private lockDir: string;

    constructor(projectRoot: string, storeSubdir = '.ai_runtime') {
        this.projectRoot = path.resolve(projectRoot);
        this.storeDir = path.join(this.projectRoot, storeSubdir);
        this.archiveDir = path.join(this.storeDir, 'snapshots');
        this.indexPath = path.join(this.storeDir, 'snapshots.json');
        this.lockDir = this.storeDir; 

        fs.mkdirSync(this.archiveDir, { recursive: true });
        if (!fs.existsSync(this.indexPath)) {
            fs.writeFileSync(this.indexPath, '[]\n');
        }
    }

    private async withLock<T>(fn: () => Promise<T>): Promise<T> {
        fs.mkdirSync(this.lockDir, { recursive: true });
        let release: () => Promise<void>;
        try {
            release = await properLockfile.lock(this.lockDir, {
                lockfilePath: path.join(this.lockDir, 'snapshot.lock.dir'),
                retries: {
                    retries: 200, // 10s total wait
                    minTimeout: 50,
                    maxTimeout: 50
                }
            });
        } catch (e: any) {
            throw new Error(`Could not acquire snapshot lock within 10s: ${e.message}`);
        }

        try {
            return await fn();
        } finally {
            await release();
        }
    }

    private readIndex(): SnapshotRecord[] {
        if (!fs.existsSync(this.indexPath)) {
            return [];
        }
        const raw = fs.readFileSync(this.indexPath, 'utf8').trim();
        if (!raw) {
            return [];
        }
        const parsed = JSON.parse(raw);
        if (!Array.isArray(parsed)) {
            throw new Error(`Snapshot index is corrupt: ${this.indexPath}`);
        }
        return parsed as SnapshotRecord[];
    }

    private writeIndex(records: SnapshotRecord[]): void {
        const tmpPath = `${this.indexPath}.tmp`;
        fs.writeFileSync(tmpPath, `${JSON.stringify(records, null, 2)}\n`);
        fs.renameSync(tmpPath, this.indexPath);
    }

    public async capture(patchId = ''): Promise<SnapshotHandle> {
        const snapId = crypto.randomUUID();
        const archivePath = path.join(this.archiveDir, `${snapId}.tar.gz`);

        const handle: SnapshotHandle = {
            snapshot_id: snapId,
            project_root: this.projectRoot,
            patch_id: patchId,
            path: archivePath,
            status: 'pending',
            created_at: Date.now() / 1000,
            archive_size_bytes: 0,
            file_count: 0
        };

        await this.withLock(async () => {
            const { fileCount, sizeBytes, relFiles } = await this.writeArchive(archivePath);
            handle.file_count = fileCount;
            handle.archive_size_bytes = sizeBytes;

            const records = this.readIndex();
            records.push({
                ...handle,
                file_manifest: relFiles
            });
            this.writeIndex(records);
        });

        return handle;
    }

    private async writeArchive(dest: string): Promise<{ fileCount: number, sizeBytes: number, relFiles: string[] }> {
        const tmpPath = `${dest}.tmp`;
        const files = collectSourceFiles(this.projectRoot);
        const relFiles = files.map(f => path.relative(this.projectRoot, f));
        
        try {
            if (relFiles.length === 0) {
                await tar.c({
                    gzip: { level: 6 },
                    file: tmpPath,
                    cwd: this.projectRoot,
                    filter: () => false
                }, ['.']);
            } else {
                await tar.c({
                    gzip: { level: 6 },
                    file: tmpPath,
                    cwd: this.projectRoot,
                }, relFiles);
            }

            fs.renameSync(tmpPath, dest);
        } catch (e) {
            if (fs.existsSync(tmpPath)) {
                fs.unlinkSync(tmpPath);
            }
            throw e;
        }

        const sizeBytes = fs.existsSync(dest) ? fs.statSync(dest).size : 0;
        return { fileCount: files.length, sizeBytes, relFiles };
    }

    public async restore(handle: SnapshotHandle): Promise<void> {
        if (!handle.path || !fs.existsSync(handle.path)) {
            throw new Error(`Snapshot archive not found: '${handle.path}'. It may have been pruned or the snapshot was never fully committed.`);
        }

        await this.withLock(async () => {
            // Load the file manifest so we know what was in the snapshot
            let manifestFiles: Set<string> = new Set();
            const records = this.readIndex();
            const record = records.find(r => r.snapshot_id === handle.snapshot_id);
            if (record?.file_manifest) {
                for (const f of record.file_manifest) {
                    // Normalise to absolute paths for comparison
                    manifestFiles.add(path.resolve(this.projectRoot, f));
                }
            }

            // Extract the archive (restores original file contents)
            await this.extractArchive(handle.path);

            // Delete files that exist NOW but were NOT in the snapshot
            if (manifestFiles.size > 0) {
                const currentFiles = collectSourceFiles(this.projectRoot);
                for (const absFile of currentFiles) {
                    if (!manifestFiles.has(absFile)) {
                        try { fs.unlinkSync(absFile); } catch { /* best-effort */ }
                    }
                }
            }

            for (const item of records) {
                if (item.snapshot_id === handle.snapshot_id) {
                    item.status = 'rolled_back';
                    break;
                }
            }
            this.writeIndex(records);
        });

        handle.status = 'rolled_back';
    }

    private async extractArchive(archivePath: string): Promise<void> {
        await tar.x({
            file: archivePath,
            cwd: this.projectRoot
        });
    }

    public commit(handle: SnapshotHandle): void {
        const records = this.readIndex();
        for (const item of records) {
            if (item.snapshot_id === handle.snapshot_id) {
                item.status = 'committed';
                break;
            }
        }
        this.writeIndex(records);
        handle.status = 'committed';
    }

    public listSnapshots(limit = 50): SnapshotHandle[] {
        return this.readIndex()
            .sort((a, b) => b.created_at - a.created_at)
            .slice(0, limit)
            .map(({ file_manifest: _fileManifest, ...handle }) => handle);
    }

    public async prune(keep = 10): Promise<number> {
        return this.withLock(async () => {
            const records = this.readIndex();
            const pruneable = records
                .filter(item => item.status === 'committed' || item.status === 'rolled_back')
                .sort((a, b) => b.created_at - a.created_at);
            const toDelete = new Set(pruneable.slice(keep).map(item => item.snapshot_id));

            let deleted = 0;
            for (const item of records) {
                if (!toDelete.has(item.snapshot_id)) {
                    continue;
                }
                try {
                    if (fs.existsSync(item.path)) {
                        fs.unlinkSync(item.path);
                    }
                } catch (e) {}
                deleted++;
            }

            this.writeIndex(records.filter(item => !toDelete.has(item.snapshot_id)));
            return deleted;
        });
    }

    public load(snapshotId: string): SnapshotHandle | null {
        const row = this.readIndex().find(item => item.snapshot_id === snapshotId);
        if (!row) return null;
        const { file_manifest: _fileManifest, ...handle } = row;
        return handle;
    }
}
